source/common/stats/stat_merger.cc

```diff
@@ -1,17 +1,84 @@
 #include "common/stats/stat_merger.h"
 
+#include <algorithm>
+
 namespace Envoy {
 namespace Stats {
 
-StatMerger::StatMerger(Stats::Store& target_store) : temp_scope_(target_store.createScope("")) {}
+StatMerger::StatMerger(Store& target_store) : temp_scope_(target_store.createScope("")) {}
+
+StatMerger::DynamicSpans StatMerger::DynamicContext::encodeComponents(StatName stat_name) {
+  DynamicSpans dynamic_spans;
+  uint32_t index = 0;
+  auto record_dynamic = [&dynamic_spans, &index](absl::string_view str) {
+    DynamicSpan span;
+    span.first = index;
+    index += std::count(str.begin(), str.end(), '.');
+    span.second = index;
+    ++index;
+    dynamic_spans.push_back(span);
+  };
+  SymbolTableImpl::Encoding::decodeTokens(
+      stat_name.data(), stat_name.dataSize(), [&index](Symbol) { ++index; }, record_dynamic);
+  return dynamic_spans;
+}
+
+StatName StatMerger::DynamicContext::makeDynamicStatName(const std::string& name,
+                                                         const DynamicsMap& map) {
+  auto iter = map.find(name);
+  if (iter == map.end()) {
+    return symbolic_pool_.add(name);
+  }
+
+  const DynamicSpans& dynamic_spans = iter->second;
+  auto dynamic = dynamic_spans.begin();
+  auto dynamic_end = dynamic_spans.end();
+
+  // Name has embedded dynamic components; we'll need to join together the
+  // static/dynamic StatName components.
+  std::vector<StatName> components;
+  uint32_t component_index = 0;
+  std::vector<absl::string_view> dynamic_tokens;
+
+  for (auto segment : absl::StrSplit(name, '.')) {
```

Comment:
Do we do anything to prevent stats that have double periods (`..`)? It seems like that could get us into trouble if we don't handle that well?